node/ttl_heap.js

```diff
@@ -0,0 +1,255 @@
+ * Overview:
+ * - ttlHeap.items and ttlHeap.ttls are the items on the heap and their time
+ *   left respectively
+ *
+ * - ttlHeap.lastTime is the timestamp (ms) that all ttls in the heap are relative to
+ *
+ * - ttlHeap.timer is the currently pending timer object or null
+ *
+ * - ttlHeap.expired are any items pending having their .onTimeout called; this
+ *   call is defered until next tick after a drain to avoid any heap
+ *   interactions/changes while draining (if the callback adds something else
+ *   to the heap)
+ *
+ * The primary public API is:
+ *     ttlHeap.update(item, ?now)
+ *
+ * Note the method is is intentionally not named something simple like "add" or
+ * "append" since it does a bit more:
+ * - update the current time, decrementing all ttls by any time elapsed
+ * - pop any expired ttls, and defer a call to their .onTimeout
+ * - push the new item and its initial ttl onto the heap
+ * - set a timer if there is none or the newly added item is the next to expire
+ *
+ * TODO: evaluate many things, including:
+ * - is it worth it to store the ttls in a typed array?
+ * - would be better off storing item and ttl in an array of records?
+ */
+
+function TTLHeap(options) {
+    var self = this;
+    options = options || {};
+
+    self.timers = options.timers || globalTimers;
+    self.minTimeout = options.minTimeout || null;
+    self.items = [];
+    self.ttls = [];
+    self.expired = [];
+    self.lastTime = self.timers.now();
+    self.timer = null;
+}
+
+TTLHeap.prototype.clear = function clear() {
+    var self = this;
+
+    self.timers.clearTimeout(self.timer);
+    self.items = [];
+    self.ttls = [];
+    self.lastTime = self.timers.now();
+    self.timer = null;
+};
+
+TTLHeap.prototype.getNextTimeout = function getNextTimeout(now) {
+    var self = this;
+
+    var timeout = self.ttls[0];
+    if (typeof self.minTimeout === 'function') {
+        timeout = Math.max(self.minTimeout(now), timeout);
+    } else if (typeof self.minTimeout === 'number') {
+        timeout = Math.max(self.minTimeout, timeout);
+    }
+    return timeout;
+};
+
+TTLHeap.prototype.update = function update(item, now) {
+    var self = this;
+
+    if (now === undefined) {
+        now = self.timers.now();
+    }
+
+    self.drainExpired(now);
+    var i = self.push(item, item.timeout);
+    // update timer if none, or the newly added item is the new root
+    if (!self.timer || i === 0) {
+        self.setNextTimer();
+    }
+};
+
+TTLHeap.prototype.setNextTimer = function setNextTimer(now) {
+    var self = this;
+
+    if (self.timer) {
+        self.timers.clearTimeout(self.timer);
+    }
+
+    var timeout = self.getNextTimeout(now);
+    self.timer = self.timers.setTimeout(onTimeout, timeout);
+
+    function onTimeout() {
+        var now = self.timers.now();
+        self.onTimeout(now);
+    }
+};
+
+TTLHeap.prototype.onTimeout = function onTimeout(now) {
+    var self = this;
+
+    self.timer = null;
+    self.drainExpired(now);
```

Comment:
Unnecessary O(n) operation if you reframe ttl absolute instead of relative time.